p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,398 @@
+			case <-d.ctx.Done():
+				return 0, io.ErrClosedPipe
+			case <-deadlineUpdated:
+
+			}
+		} else {
+			return d.writeMessage(msg)
+		}
+	}
+}
+
+func (d *dataChannel) writeMessage(msg *pb.Message) (int, error) {
+	err := d.writer.WriteMsg(msg)
+	// this only returns the number of bytes sent from the buffer
+	// requested by the user.
+	return len(msg.GetMessage()), err
+
+}
+
+func (d *dataChannel) Close() error {
+	select {
+	case <-d.ctx.Done():
+		return nil
+	default:
+	}
+
+	d.m.Lock()
+	d.state = stateClosed
+	d.m.Unlock()
+
+	d.cancel()
+	d.CloseWrite()
+	_ = d.channel.Close()
+	d.wg.Wait()
+	return nil
+}
+
+func (d *dataChannel) CloseRead() error {
+	var err error
+	d.closeReadOnce.Do(func() {
+		d.m.Lock()
+		previousState := d.state
+		currentState := d.state.processOutgoingFlag(pb.Message_STOP_SENDING)
+		d.state = currentState
+		d.m.Unlock()
+		if previousState != currentState && currentState == stateClosed {
+			defer d.Close()
+		}
+		msg := &pb.Message{
+			Flag: pb.Message_STOP_SENDING.Enum(),
+		}
+		err = d.writer.WriteMsg(msg)
+	})
+	return err
+
+}
+
+func (d *dataChannel) remoteClosed() {
+	d.m.Lock()
+	defer d.m.Unlock()
+	d.state = stateClosed
+	d.cancel()
+
+}
+
+func (d *dataChannel) CloseWrite() error {
+	var err error
+	d.closeWriteOnce.Do(func() {
+		d.m.Lock()
+		previousState := d.state
+		currentState := d.state.processOutgoingFlag(pb.Message_FIN)
+		d.state = currentState
+		d.m.Unlock()
+		if previousState != currentState && currentState == stateClosed {
+			defer d.Close()
+		}
+		msg := &pb.Message{
+			Flag: pb.Message_FIN.Enum(),
+		}
+		err = d.writer.WriteMsg(msg)
+	})
+	return err
+}
+
+func (d *dataChannel) LocalAddr() net.Addr {
+	return d.laddr
+}
+
+func (d *dataChannel) RemoteAddr() net.Addr {
+	return d.raddr
+}
+
+func (d *dataChannel) Reset() error {
+	var err error
+	d.resetOnce.Do(func() {
+		msg := &pb.Message{Flag: pb.Message_RESET.Enum()}
+		_, err = d.writeMessage(msg)
+		d.Close()
+	})
+	return err
+}
+
+func (d *dataChannel) SetDeadline(t time.Time) error {
+	d.m.Lock()
+	defer d.m.Unlock()
+	d.readDeadline = t
+	d.writeDeadline = t
+	return nil
+}
+
+func (d *dataChannel) SetReadDeadline(t time.Time) error {
+	d.m.Lock()
+	defer d.m.Unlock()
+	d.readDeadline = t
+	deadlineUpdated := d.deadlineUpdated
+	d.deadlineUpdated = make(chan struct{})
+	defer func() { close(deadlineUpdated) }()
+	return nil
+}
+
+func (d *dataChannel) SetWriteDeadline(t time.Time) error {
+	d.m.Lock()
+	defer d.m.Unlock()
+	d.writeDeadline = t
+	deadlineUpdated := d.deadlineUpdated
+	d.deadlineUpdated = make(chan struct{})
+	defer func() { close(deadlineUpdated) }()
+	return nil
+}
+
+func (d *dataChannel) getState() channelState {
+	d.m.Lock()
+	defer d.m.Unlock()
+	return d.state
+}
+
+func (d *dataChannel) readLoop() {
+	defer d.wg.Done()
+	for {
+		select {
+		case <-d.ctx.Done():
+			return
+		case <-d.requestRead:
+		}
+
+		var msg pb.Message
+		err := d.reader.ReadMsg(&msg)
```

Comment:
I think you leak the goroutine here because I don't see any timeout on `d.reader` (so if someone try to close it `d.ctx.Done()` would be done but you could block on `ReadMsg` forever)

Reply:
There is no way to currently add a direct deadline on the reader as it is not supported by pion. https://github.com/pion/datachannel/issues/157 . This is a good test case. It shouldn't leak since the context only closes when the underlying datachannel closes. That should make the Read call return as `io.EOF`.

Reply:
ok I see, thx